include/net/net_pkt.h

```diff
@@ -93,6 +93,13 @@ struct net_pkt {
 
 	/** @cond ignore */
 
+#if defined(CONFIG_WIFI_ESP)
+	/* TODO: The work is to be removed after the esp offloading
+	 * wifi driver is modified to use its own private work struct.
+	 */
+	struct k_work esp_work;
+#endif
```

Comment:
I need to find some time to fix/change ESP driver, it won't happen this week. I wonder if there might be other downstream drivers that use `pkt->work`. `net_pkt_work()` API was kind of public, so I wonder if we should follow deprecation policy here for other drivers. So for example instead of introducing `esp_work` and removing `work`, add something like:
```
struct net_pkt {
	union {
#ifdef CONFIG_NET_PKT_WORK
		/** Internal variable that is used when packet is sent
		 * or received.
		 */
		struct k_work work;
#endif
...
```
with hidden Kconfig option:
```
config NET_PKT_WORK
	bool
```
which could be selected by ESP driver and any other downstream drivers or L2 layers.

Reply:
The `k_work` is strictly for internal use so we are free to remove it if needed. Any downstream user using it will need to change its code.

Reply:
Then what makes `struct k_work work;` different from `struct net_pkt_cursor cursor;`? They are both described as `Internal variable...` or `Internal buffer...`, which means that neither should be used by downstream code, right? But if someone wants to implement driver like ESP-AT (but for different chip), then `net_pkt_cursor_init()` still needs to be used for properly initializing packet before pushing it to other networking layers. This however makes `pkt->cursor` no longer internal. Well, it all depends on what "internal" really means.

Maybe we should better describe which structure members are reserved by native networking stack? And prevent using them by networking drivers, like ESP-AT does right now with `pkt->work`?

Reply:
> Maybe we should better describe which structure members are reserved by native networking stack? And prevent using them by networking drivers, like ESP-AT does right now with `pkt->work`?

Yes, that would be a good thing to do. It is a bit difficult to prevent the code from using the "internal" variables as we need to have net_pkt in the "open". The access functions for the internal variables are not documented so any downstream code should not use these vars anyway. For your cursor example, the cursor manipulation API is there and documented so I would not consider it as internal. The apps should never access the variables directly anyway.